lib/Containers/MerkleTree.cpp

```diff
@@ -910,36 +1146,95 @@ template <typename Hasher, std::uint64_t const BranchingBits>
 typename MerkleTree<Hasher, BranchingBits>::Meta&
 MerkleTree<Hasher, BranchingBits>::meta() noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(_buffer);
-  return *reinterpret_cast<Meta*>(_buffer.get());
+  return _data.meta;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
-typename MerkleTree<Hasher, BranchingBits>::Meta const&
-MerkleTree<Hasher, BranchingBits>::meta() const noexcept {
+MerkleTreeBase::Meta const& MerkleTree<Hasher, BranchingBits>::meta() const noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(_buffer);
-  return *reinterpret_cast<Meta const*>(_buffer.get());
+  return _data.meta;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
-typename MerkleTree<Hasher, BranchingBits>::Node&
-MerkleTree<Hasher, BranchingBits>::node(std::uint64_t index) noexcept {
+MerkleTreeBase::Node& MerkleTree<Hasher, BranchingBits>::node(std::uint64_t index) {
   // not thread-safe, lock buffer from outside
+  
+  // determine shard
+  std::uint64_t shard = shardForIndex(index);
+  TRI_ASSERT(shard < numberOfShards());
+
+  // lazily allocate backing memory
+  if (_data.shards.size() <= shard) {
+    // allocate a few shards at once
+    _data.shards.resize(std::max(shard + 1, std::uint64_t(4)));
+  }
+  if (_data.shards[shard] == nullptr) {
+    _data.shards[shard] = std::make_unique<uint8_t[]>(ShardSize);
+    Node* p = reinterpret_cast<Node*>(_data.shards[shard].get());
+    
+    std::uint64_t nodesPerShard = shardSize(meta().depth) / NodeSize;
+    for (std::uint64_t i = 0; i < nodesPerShard; ++i) {
```

Comment:
This could potentially also be done by `memset` and could be a performance critical part...

Reply:
Done